### PR TITLE
drivers: sensors: sm351lt: Add new driver

### PR DESCRIPTION
**Pull request by nordicjm**
Signed-off-by: Jamie McCrae <jamie.mccrae@lairdconnect.com>

### CODE REVIEW

File: drivers/sensor/sm351lt/sm351lt.c [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+		val->val2 = 0;
+	} else {
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+#if CONFIG_SM351LT_TRIGGER
+int sm351lt_attr_set(struct device *dev, enum sensor_channel chan,
```
Comment (MaureenHelm) [inline drivers/sensor/sm351lt/sm351lt.c:135]: Make this function `static`

File: drivers/sensor/sm351lt/sm351lt.c [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+			return -ENOTSUP;
+		}
+	} else {
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+int sm351lt_attr_get(struct device *dev, enum sensor_channel chan,
```
Comment (MaureenHelm) [inline drivers/sensor/sm351lt/sm351lt.c:154]: Make this function `static`

File: samples/sensor/sm351lt/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#
+# Copyright (c) 2020, Laird Connectivity
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+cmake_minimum_required(VERSION 3.13.1)
```
Comment (MaureenHelm) [inline samples/sensor/sm351lt/CMakeLists.txt:7]: Please move the sample to a separate commit

File: samples/sensor/sm351lt/README.rst [hunk 4]
```diff
@@ -0,0 +1,50 @@
+.. _sm351lt:
+
+SM351LT: Magnetoresistive Sensor Example
+########################################
+
+Overview
+********
+
+This sample application periodically polls an SM351LT magnetoresistive sensor
+and displays if a magnet near to the sensor or not, via the console.
```
Comment (MaureenHelm) [inline samples/sensor/sm351lt/README.rst:10]:
```suggestion
and displays if a magnet is near to the sensor, via the console.
```

File: samples/sensor/sm351lt/sample.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+sample:
+  name: SM351LT Magnetoresistive Sensor Example
+tests:
+  sample.sensor.sm351lt:
+    harness: sensor
+    tags: samples sensor
+    depends_on: gpio sm351lt
```
Comment (MaureenHelm) [inline samples/sensor/sm351lt/sample.yaml:7]: Does the bt510 board (or any board) have `sm3511t` in its supported list? If not, this sample will not get built in CI

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: samples/sensor/sm351lt/sample.yaml (hunk 5 above)
Comment (lairdjm) [reply samples/sensor/sm351lt/sample.yaml:7]: The bt510 board is pending review here: https://github.com/zephyrproject-rtos/zephyr/pull/27400
But the dts doesn't have the sm351lt in it at present, the idea was to get the basic board added alongside this, then get the bt510 board updated to include this otherwise I'm assuming the board would fail submission because the automated checks would fail and say sm351lt not a known driver, or can we add it prior to the driver existing?

File: drivers/sensor/sm351lt/sm351lt.h [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+#include <drivers/gpio.h>
+#include <drivers/sensor.h>
+#include <string.h>
+
+#define SENSOR_ATTR_SM351LT_TRIGGER_TYPE SENSOR_ATTR_PRIV_START
+
+struct sm351lt_config {
+	char *bus_name;
+	uint8_t	gpio;
+	uint8_t	gpio_flags;
```
Comment (gmarull) [inline drivers/sensor/sm351lt/sm351lt.h:23]: should be `gpio_dt_flags_t`

File: drivers/sensor/sm351lt/sm351lt.h [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+#include <stdint.h>
+#include <drivers/gpio.h>
+#include <drivers/sensor.h>
+#include <string.h>
+
+#define SENSOR_ATTR_SM351LT_TRIGGER_TYPE SENSOR_ATTR_PRIV_START
+
+struct sm351lt_config {
+	char *bus_name;
+	uint8_t	gpio;
```
Comment (gmarull) [inline drivers/sensor/sm351lt/sm351lt.h:22]: should be `gpio_pin_t`, name should probably be adjusted too.

File: drivers/sensor/sm351lt/sm351lt.h [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+#include <sys/util.h>
+#include <stdint.h>
+#include <drivers/gpio.h>
+#include <drivers/sensor.h>
+#include <string.h>
+
+#define SENSOR_ATTR_SM351LT_TRIGGER_TYPE SENSOR_ATTR_PRIV_START
+
+struct sm351lt_config {
+	char *bus_name;
```
Comment (gmarull) [inline drivers/sensor/sm351lt/sm351lt.h:21]: should be `const char *`

File: drivers/sensor/sm351lt/sm351lt.h [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+#ifndef ZEPHYR_DRIVERS_SENSOR_SM351LT_SM351LT_H_
+#define ZEPHYR_DRIVERS_SENSOR_SM351LT_SM351LT_H_
+
+#include <kernel.h>
+#include <device.h>
+#include <sys/util.h>
+#include <stdint.h>
+#include <drivers/gpio.h>
+#include <drivers/sensor.h>
+#include <string.h>
```
Comment (gmarull) [inline drivers/sensor/sm351lt/sm351lt.h:16]: Header files should only include necessary stuff

File: drivers/sensor/sm351lt/sm351lt.c [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+#define DT_DRV_COMPAT honeywell_sm351lt
+
+#include <init.h>
+#include <sys/byteorder.h>
+#include <sys/__assert.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(SM351LT, CONFIG_SENSOR_LOG_LEVEL);
+
+#include "sm351lt.h"
```
Comment (gmarull) [inline drivers/sensor/sm351lt/sm351lt.c:16]: This include should go to the top (first)

File: drivers/sensor/sm351lt/sm351lt.c [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+	struct sm351lt_data *data = dev->driver_data;
+	int ret = -ENOTSUP;
+
+	if (trig->chan == SENSOR_CHAN_PROX) {
+		data->changed_handler = handler;
+		ret = gpio_pin_interrupt_configure(data->bus, config->gpio,
+						   (handler ? data->trigger_type
+							    : GPIO_INT_DISABLE));
+
+		if (ret != 0) {
```
Comment (gmarull) [inline drivers/sensor/sm351lt/sm351lt.c:33]: `ret < 0`

File: drivers/sensor/sm351lt/sm351lt.c [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+#endif
+#endif
+
+static int sm351lt_sample_fetch(struct device *dev,
+				enum sensor_channel chan)
+{
+	const struct sm351lt_config *config = dev->config_info;
+	struct sm351lt_data *data = dev->driver_data;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL || chan == SENSOR_CHAN_PROX);
```
Comment (gmarull) [inline drivers/sensor/sm351lt/sm351lt.c:111]: return `-ENOTSUP` if requested channel is not supported.

File: drivers/sensor/sm351lt/sm351lt.c [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+
+static int sm351lt_init(struct device *dev)
+{
+	const struct sm351lt_config *const config = dev->config_info;
+	struct sm351lt_data *data = dev->driver_data;
+	uint32_t ret;
+
+	data->bus = device_get_binding(config->bus_name);
+	if (!data->bus) {
+		LOG_DBG("gpio bus not found: %s", config->bus_name);
```
Comment (gmarull) [inline drivers/sensor/sm351lt/sm351lt.c:195]: `LOG_ERR`

File: drivers/sensor/sm351lt/sm351lt.c [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+		return -EINVAL;
```
Comment (gmarull) [inline drivers/sensor/sm351lt/sm351lt.c:196]: `-ENODEV`

File: drivers/sensor/sm351lt/sm351lt.c [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+	}
+
+	ret = gpio_pin_configure(data->bus, config->gpio,
+				 (config->gpio_flags | GPIO_INPUT));
+	if (ret) {
+		LOG_DBG("failed to configure gpio: %d", ret);
```
Comment (gmarull) [inline drivers/sensor/sm351lt/sm351lt.c:202]: `LOG_ERR`

File: drivers/sensor/sm351lt/sm351lt.c [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+	data->work.handler = sm351lt_work_cb;
+	data->dev = dev;
+#endif
+
+	data->trigger_type = GPIO_INT_DISABLE;
+
+	ret = gpio_pin_interrupt_configure(data->bus, config->gpio,
+					   GPIO_INT_DISABLE);
+	if (ret) {
+		LOG_DBG("failed to configure gpio interrupt: %d", ret);
```
Comment (gmarull) [inline drivers/sensor/sm351lt/sm351lt.c:234]: `LOG_ERR`

File: drivers/sensor/sm351lt/sm351lt.c [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+
+	/* Setup callback struct but do not add it yet */
+	gpio_init_callback(&data->gpio_cb, sm351lt_gpio_callback,
+			   BIT(config->gpio));
+#endif
+
+	return 0;
+}
+
+/* Instantiation macros	for each individual device. */
```
Comment (gmarull) [inline drivers/sensor/sm351lt/sm351lt.c:246]: extra spaces should be removed

**Review by gmarull — CHANGES_REQUESTED** [review]

**Review by rerickson1 — APPROVED** [review]

File: drivers/sensor/sm351lt/sm351lt.c [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+#endif
+#endif
+
+static int sm351lt_sample_fetch(struct device *dev,
+				enum sensor_channel chan)
+{
+	const struct sm351lt_config *config = dev->config;
+	struct sm351lt_data *data = dev->data;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL || chan == SENSOR_CHAN_PROX);
```
Comment (gmarull) [inline drivers/sensor/sm351lt/sm351lt.c:115]: This assertion is now redundant.

**Review by gmarull — APPROVED** [review]

File: drivers/sensor/sm351lt/sm351lt.c [hunk 19]
```diff
@@ -0,0 +1,271 @@
...
+			(k_thread_entry_t)sm351lt_thread, dev, NULL,
+			NULL, K_PRIO_COOP(CONFIG_SM351LT_THREAD_PRIORITY),
+			0, K_NO_WAIT);
+
+#if defined(CONFIG_THREAD_NAME) && defined(CONFIG_THREAD_MAX_NAME_LEN)
+	/* Sets up thread name based on the device name */
+	char thread_name[CONFIG_THREAD_MAX_NAME_LEN];
+	memset(thread_name, 0, CONFIG_THREAD_MAX_NAME_LEN);
+	snprintf(thread_name, CONFIG_THREAD_MAX_NAME_LEN, "sen_%s", dev->name);
+	k_thread_name_set(&data->thread, thread_name);
```
Comment (MaureenHelm) [inline drivers/sensor/sm351lt/sm351lt.c:226]: Why not just use the dev name here and skip the runtime string manipulation?

File: drivers/sensor/sm351lt/sm351lt.c [hunk 20]
```diff
@@ -0,0 +1,271 @@
...
+#include <stdio.h>
+#include <sys/util.h>
+#include <drivers/gpio.h>
+#include <drivers/sensor.h>
+#include <string.h>
+
+LOG_MODULE_REGISTER(SM351LT, CONFIG_SENSOR_LOG_LEVEL);
+
+#if CONFIG_SM351LT_TRIGGER
+int sm351lt_trigger_set(struct device *dev,
```
Comment (MaureenHelm) [inline drivers/sensor/sm351lt/sm351lt.c:23]: This should be `static`

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please have a look at the buildkite failure

File: drivers/sensor/sm351lt/sm351lt.c (hunk 19 above)
Comment (lairdjm) [reply drivers/sensor/sm351lt/sm351lt.c:226]: For debugging purposes, it prefixes the thread name with 'sen_' to indicate sensor

**Comment by lairdjm** [comment]
@MaureenHelm The build failure was caused by an STM32HAL error (nothing to do with my changes) and the second part is being caused by something else, namely https://github.com/zephyrproject-rtos/zephyr/pull/27598 for some reason this was approved and merged in but has been merged in with the wrong commit and so an old broken commit has been merged in somehow.

**Review by MaureenHelm — APPROVED** [review]
